File: Go/Algorithms/Sorting/quickSort.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)


func main()  {
	fmt.Println("Please input numbers(separate with space):")
	br := bufio.NewReader(os.Stdin)
	a, _, _ := br.ReadLine()
	ns := strings.Split(string(a), " ")
	var values []int
	for _, s := range(ns) {
		n, _ := strconv.Atoi(s)
		values = append(values, n)
	}
	QuickSort(values)
	fmt.Println(values)
}

func QuickSort(arr []int) []int {
	return QuickSortUtil(arr, 0, len(arr) - 1);
}

func QuickSortUtil(arr []int, low, high int) []int {
	if low < high {
		var p int
		arr, p = Partition(arr, low, high)
		arr = QuickSortUtil(arr, low, p-1)
		arr = QuickSortUtil(arr, p+1, high)
	}
	return arr
}

func Partition(arr []int, low, high int) ([]int, int) {
	pivot := arr[high]
	i := low
	for j := low; j < high; j++ {
		if arr[j] < pivot {
			arr[i], arr[j] = arr[j], arr[i]
			i++
		}
	}
	arr[i], arr[high] = arr[high], arr[i]
	return arr, i
}

File: Python/Algorithms/Sorting/quicksort.py

def partition(arr, low, high):
    index_smaller = (low - 1)
    pivot = arr[high]

    for j in range(low, high):
        if arr[j] <= pivot:
            index_smaller += 1
            arr[index_smaller], arr[j] = arr[j], arr[index_smaller]

    arr[index_smaller + 1], arr[high] = arr[high], arr[index_smaller + 1]

    return index_smaller + 1


def quicksort_iteractive(arr, start, end):
    size = (end - start) + 1 
    stack = [0] * size

    top = -1

    top += 1
    stack[top] = start
    top += 1
    stack[top] = end

    while top >= 0:
        end = stack[top]
        top -= 1
        start = stack[top]
        top -= 1

        p = partition(arr, start, end)

        if p - 1 > start:
            top += 1
            stack[top] = start
            top += 1
            stack[top] = p - 1

        if p + 1 < end:
            top += 1
            stack[top] = p + 1
            top += 1
            stack[top] = end

def quicksort_recursive(arr, start, end):
    if start < end:
        p = partition(arr, start, end)

        quicksort_recursive(arr, start, p-1)
        quicksort_recursive(arr, p+1, end)


